File: BEmuJava/bemu/src/main/java/com/bloomberglp/blpapi/ReferenceElementArrayChainTickersItem.java
//------------------------------------------------------------------------------
// <copyright project="BEmu_maven" file="/BEmu_maven/bemu/src/main/java/com/bloomberglp/blpapi/ReferenceElementArrayChainTickersItem.java" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

package com.bloomberglp.blpapi;

import java.text.DateFormat;
import java.text.SimpleDateFormat;

import com.bloomberglp.blpapi.Datetime;
import com.bloomberglp.blpapi.Name;
import com.bloomberglp.blpapi.SchemaTypeDefinition;

public class ReferenceElementArrayChainTickersItem extends Element
{
    private final ReferenceElementString _element;

    ReferenceElementArrayChainTickersItem(String ticker, Datetime dtExp, OptionalityEnum optionality, int strike)
    {
    	DateFormat df = new SimpleDateFormat ("MM/dd/yy");
        String optionTicker = String.format("%s US %s %s%s", 
        		ticker, 
        		df.format(dtExp.calendar().getTime()), 
    			optionality.toString().toUpperCase(), 
    			String.valueOf(strike));
        
        this._element = new ReferenceElementString("Ticker", optionTicker);
    }
    
    public SchemaTypeDefinition typeDefinition() throws Exception
    {
    	return new SchemaTypeDefinition(this.datatype(), new Name("CHAIN_TICKERS"));
    }
    
    public Name name() throws Exception
    {
    	return new Name("Ticker");
    }
    
    public int numValues()
    {
    	return 0;
    }
    
    public int numElements()
    {
    	return 1;
    }
    
    public boolean isComplexType()
    {
    	return false;
    }
    
    public boolean isArray()
    {
    	return false;
    }
    
    public boolean isNull()
    {
    	return false;
    }
    
    public String getElementAsString(String name) throws Exception
    {
    	if(this._element.name().toString().equals(name))
    		return this._element.getValueAsString();
    	
    	else
    		return super.getElementAsString(name);
    }
    
    protected StringBuilder prettyPrint(int tabIndent) throws Exception
    {
        String tabs = com.bloomberglp.blpapi.IndentType.Indent(tabIndent);
        StringBuilder result = new StringBuilder();

        result.append(String.format("%sCHAIN_TICKERS = {%s", tabs, System.getProperty("line.separator")));
        result.append(this._element.prettyPrint(tabIndent + 1));
        result.append(String.format("%s}%s", tabs, System.getProperty("line.separator")));
        
        return result;
    }
    
    
}

File: BEmu/ReferenceDataRequest/ReferenceElementArrayChainTickersItem.cs
//------------------------------------------------------------------------------
// <copyright project="BEmu_csh" file="ReferenceDataRequest/ReferenceElementArrayChainTickersItem.cs" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

namespace Bloomberglp.Blpapi.ReferenceDataRequest
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Text;

    internal class ReferenceElementArrayChainTickersItem : Element
    {
        private readonly ReferenceElementString _element;

        internal ReferenceElementArrayChainTickersItem(string ticker, DateTime dtExp, ReferenceElementArrayChainTickers.OptionalityEnum optionality, int strike)
        {
            string optionTicker = string.Format("{0} US {1:MM/dd/yy} {2}{3}", ticker, dtExp, (char)optionality, strike);
            this._element = new ReferenceElementString("Ticker", optionTicker);
        }

        public override Name Name { get { return new Name("Ticker"); } }
        public override int NumValues { get { return 0; } }
        public override int NumElements { get { return 1; } }
        public override bool IsArray { get { return false; } }
        public override bool IsComplexType { get { return false; } }
        public override IEnumerable<Element> Elements { get { yield return this._element; } }
        public override string GetElementAsString(string name) { return this[name].GetValueAsString(); }
        public override SchemaTypeDefinition TypeDefinition { get { return new SchemaTypeDefinition(this.Datatype, new Name("CHAIN_TICKERS")); } }

        //public override string GetElementAsString(string name)
        public override Element this[string name]
        {
            get
            {
                foreach (var item in this.Elements)
                {
                    if (item.Name.ToString() == name)
                        return item;
                }
                return base[name];
            }
        }


        internal override StringBuilder PrettyPrint(int tabIndent, bool surroundValueWithQuotes = false)
        {
            string tabs = Types.IndentType.Indent(tabIndent);
            StringBuilder result = new StringBuilder();

            result.AppendFormat("{0}CHAIN_TICKERS = {{{1}", tabs, Environment.NewLine);

            result.Append(this._element.PrettyPrint(tabIndent + 1));
            
            result.AppendFormat("{0}}}{1}", tabs, Environment.NewLine);
            return result;
        }
    }
}
